lms/src/lms/public.rs

```diff
@@ -0,0 +1,302 @@
+
+use crate::error::LmsDeserializeError;
+use crate::lms::Signature;
+use crate::types::Typecode;
+use crate::{constants::D_INTR, lms::LmsMode};
+use digest::{Digest, OutputSizeUser};
+use generic_array::{ArrayLength, GenericArray};
+use signature::{Error, Verifier};
+use typenum::{Sum, U24};
+
+//use crate::signature::Signature as Signature;
+use crate::types::Identifier;
+
+use digest::Output;
+
+#[derive(Debug)]
+/// Opaque struct representing a LMS public key
+pub struct VerifyingKey<Mode: LmsMode> {
+    pub(crate) id: Identifier,
+    pub(crate) k: Output<Mode::Hasher>,
+}
+
+impl<Mode: LmsMode> Clone for VerifyingKey<Mode> {
+    fn clone(&self) -> Self {
+        Self {
+            id: self.id,
+            k: self.k.clone(),
+        }
+    }
+}
+
+impl<Mode: LmsMode> PartialEq for VerifyingKey<Mode> {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id && self.k == other.k
+    }
+}
+
+impl<Mode: LmsMode> VerifyingKey<Mode> {
+    pub fn new(id: Identifier, k: Output<Mode::Hasher>) -> Self {
+        Self { id, k }
+    }
+
+    /// Returns the 16-byte identifier of the public key
+    pub fn id(&self) -> &Identifier {
+        &self.id
+    }
+
+    /// Returns the N-byte public key as a byte slice
+    pub fn k(&self) -> &[u8] {
+        &self.k
+    }
+}
+
+impl<Mode: LmsMode> Verifier<Signature<Mode>> for VerifyingKey<Mode> {
+    fn verify(&self, msg: &[u8], signature: &Signature<Mode>) -> Result<(), Error> {
+        // Compute the LMS Public Key Candidate Tc from the signature,
+        //    message, identifier, pubtype, and ots_typecode, using
+        //    Algorithm 6a.
+        let key_candidate = signature
+            .lmots_sig
+            .recover_pubkey(self.id, signature.q, msg);
+
+        let mut node_num = signature.q + Mode::LEAVES;
+        let mut tmp = Mode::Hasher::new()
+            .chain_update(self.id)
+            .chain_update(node_num.to_be_bytes())
+            .chain_update(D_LEAF)
+            .chain_update(key_candidate.k)
+            .finalize();
+
+        for i in 0..Mode::H {
+            // Tc = H(I || u32str(node_num/2) || u16str(D_INTR) || path[i] || tmp)
+            let mut hasher = Mode::Hasher::new()
+                .chain_update(self.id)
+                .chain_update((node_num / 2).to_be_bytes())
+                .chain_update(D_INTR);
+            if node_num % 2 == 1 {
+                hasher.update(&signature.path[i]);
+                hasher.update(&tmp);
+            } else {
+                // Tc = H(I || u32str(node_num/2) || u16str(D_INTR) || tmp || path[i])
+                hasher.update(&tmp);
+                hasher.update(&signature.path[i]);
+            }
+            hasher.finalize_into(&mut tmp);
+            node_num /= 2;
+        }
+        if self.k == tmp {
+            Ok(())
+        } else {
+            Err(Error::new())
+        }
+    }
+}
+
+/// Converts a [VerifyingKey] into its byte representation
```

Comment:
To properly linkify in rustdoc, `VerifyingKey` needs to be surrounded with backticks:

```suggestion
/// Converts a [`VerifyingKey`] into its byte representation
```